proposals/0054-multiple-asset-transfer-types-instructions.md

```diff
@@ -0,0 +1,266 @@
+
+There should be no security risks related to the new instruction from the XCVM perspective. It follows the same
+pattern as with single-type asset transfers, only now it allows combining multiple types at once.
+
+_Improves_ security by enabling [enforcement of single asset for buying execution](https://github.com/paritytech/polkadot-sdk/issues/2423),
+which minimizes the potential free/unpaid work that a receiving chain has to do. It does so, by making the
+required execution fee payment part of the instruction logic through the `remote_fees: Option<AssetFilter>`
+parameter, which will make sure the remote XCM starts with a single-asset-holding-loading-instruction, immediately
+followed by a `BuyExecution` using said asset.
+
+## Impact
+
+No impact to the rest of the spec. These are new, independent instructions, no changes to existing instructions.
+
+## Alternatives
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+1. Modifying current behavior of `TransferReserveAsset`, `DepositReserveAsset`, `InitiateReserveWithdraw` to also only
+"stage" a transfer, then execute it atomically (within same XCM execution) on destination. This is undesirable
+because it would be a breaking change. Easier to create new instructions and maybe even deprecate above if desired.
+2. Another alternative would be to create a single new instruction:
+   ```rust
+   InitiateAssetsTransfers {
+       teleports: Option<AssetFilter>,
+       local_reserve_assets: Option<AssetFilter>,
+       dest_reserve_assets: Option<AssetFilter>,
+       dest: Location,
+       remote_fees: Option<AssetFilter>,
+       remote_xcm: Xcm<()>
+   }
+   ```
+   instead of the independent `TeleportTransferAssets(AssetFilter)`, `LocalReserveDepositAssets(AssetFilter)`,
+   `DestinationReserveWithdrawAssets(AssetFilter)` "staging" instructions plus the
+   `ExecuteAssetTransfers { dest: Location, remote_fees: Option<AssetFilter>, remote_xcm: Xcm<()> }` enactment
+   instructions.
+3. Create a new `Send { dest: Location, remote_xcm: Xcm<()> }` instruction to allow building of custom asset
+   transfers using existing commands: e.g. for a teleport append `WithdrawAsset` and `BurnAsset` to an XCM to
+   be locally executed and append `ReceiveTeleportedAsset` to an XCM to be sent to destination chain. 
+   One could build whatever "custom" local and remote XCMs, then append `Send { dest, remote_xcm }` to the local
+   one and execute it, which will also forward onward `remote_xcm` to be executed with source chain origin.
+   However, this looks like a dangerous option as it enables execution of arbitrary instructions on behalf of
+   the source chain and would need complex barriers in place to be made safe. Therefore, I deem it more complex
+   and error-prone.
+
+### What is the impact of not doing this?
+
+Current multi-chain transfers are forced to happen in multiple programs per asset per "hop", resulting in very poor UX.
```

Comment:
For someone not having done a lot of XCM, one can only imagine why this is a very poor UX. What problems does a developer actually face with the current design?

Reply:
The developer simply can not write an XCM program that can, for example, transfer **both** DOT (local reserve deposit) _and_ HDX (destination reserve withdraw) from AssetHub to HydraDX. It is simply not possible with the current instruction set.

The only XCM instructions available currently (XCMv4) for transferring assets are `DepositReserveAsset`, `InitiateReserveWithdraw` and `InitiateTeleport` instructions. They initiate asset transfers on execution, but they are opinionated in the type of transfer to use. Combining them is not possible, because as a result of their individual execution, a message containing a `ClearOrigin` instruction is sent to the destination chain (for security reasons - more details available in their docs), making subsequent transfers impossible.